core/src/mast/debuginfo/mod.rs

```diff
@@ -1,7 +1,258 @@
+    pub fn is_empty(&self) -> bool {
+        self.decorators.is_empty() && self.error_codes.is_empty()
+    }
+
+    /// Strips all debug information, removing decorators and error codes.
+    /// This is used for release builds where debug info is not needed.
+    pub fn strip(&mut self) {
+        self.clear_decorators();
+        self.decorators = IndexVec::new();
+        self.error_codes.clear();
+    }
+
+    // DECORATOR ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the number of decorators.
+    pub fn num_decorators(&self) -> usize {
+        self.decorators.len()
+    }
+
+    /// Returns all decorators as a slice.
+    pub fn decorators(&self) -> &[Decorator] {
+        self.decorators.as_slice()
+    }
+
+    /// Returns the decorator with the given ID, if it exists.
+    pub fn decorator(&self, decorator_id: DecoratorId) -> Option<&Decorator> {
+        self.decorators.get(decorator_id)
+    }
+
+    /// Returns the before-enter decorators for the given node.
+    pub fn before_enter_decorators(&self, node_id: MastNodeId) -> &[DecoratorId] {
+        self.node_decorator_storage.get_before_decorators(node_id)
+    }
+
+    /// Returns the after-exit decorators for the given node.
+    pub fn after_exit_decorators(&self, node_id: MastNodeId) -> &[DecoratorId] {
+        self.node_decorator_storage.get_after_decorators(node_id)
+    }
+
+    /// Returns decorators for a specific operation within a node.
+    pub fn decorators_for_operation(
+        &self,
+        node_id: MastNodeId,
+        local_op_idx: usize,
+    ) -> &[DecoratorId] {
+        self.op_decorator_storage
+            .decorator_ids_for_operation(node_id, local_op_idx)
+            .unwrap_or(&[])
+    }
+
+    /// Returns decorator links for a node, including operation indices.
+    pub fn decorator_links_for_node(
+        &self,
+        node_id: MastNodeId,
+    ) -> Result<DecoratedLinks<'_>, DecoratorIndexError> {
+        self.op_decorator_storage.decorator_links_for_node(node_id)
+    }
+
+    // DECORATOR MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds a decorator and returns its ID.
+    pub fn add_decorator(&mut self, decorator: Decorator) -> Result<DecoratorId, MastForestError> {
+        self.decorators.push(decorator).map_err(|_| MastForestError::TooManyDecorators)
+    }
+
+    /// Adds node-level decorators (before_enter and after_exit) for the given node.
+    pub fn add_node_decorators(
+        &mut self,
+        node_id: MastNodeId,
+        before_enter: &[DecoratorId],
+        after_exit: &[DecoratorId],
+    ) {
+        self.node_decorator_storage
+            .add_node_decorators(node_id, before_enter, after_exit);
+    }
+
+    /// Adds decorator info for operations within a node.
+    pub fn add_decorator_info_for_node(
```

Comment:
nit: I think this is used specifically for basic blocks - so, I'd probably rename this to something like `add_basic_block_decorators()`.

Reply:
Actually, thinking about this a bit more, maybe this should be `register_basic_block_decorators()` because we are not really adding decorators here, but registering already added decorators with a specific basic block.

Similar logic applies to `add_node_decorators()`: maybe it should be `register_node_decorators()`.

Reply:
renamed.